Tests/BigIntTests/BigIntTests.swift

```diff
@@ -0,0 +1,479 @@
+    }
+
+    var r: BigInt = 1
+    for bit in bits_of_n(rhs).reversed() {
+      r *= r
+      if bit == 1 {
+        r *= lhs
+      }
+    }
+
+    return r
+  }
+}
+
+final class BigIntTests: XCTestCase {
+
+  #if (arch(i386) || arch(x86_64)) && !os(Windows) && !os(Android)
+  typealias FloatXX = Float80
+  #else
+  typealias FloatXX = Float64
+  #endif
+
+  /// Python: `bitWidth = 1024; -(2 ** (bitWidth - 1))`
+  static let descriptionInt1024Min: String =
+    """
+    -89884656743115795386465259539451236680898848947115328636715040578866337902\
+    750481566354238661203768010560056939935696678829394884407208311246423715319\
+    737062188883946712432742638151109800623047059726541476042502884419075341171\
+    231440736956555270413618581675255342293149119973622969239858152417678164812\
+    112068608
+    """
+
+  /// Python: `bitWidth = 1024; (2 ** (bitWidth - 1)) - 1`
+  static let descriptionInt1024Max: String =
+    """
+    89884656743115795386465259539451236680898848947115328636715040578866337902\
+    750481566354238661203768010560056939935696678829394884407208311246423715319\
+    737062188883946712432742638151109800623047059726541476042502884419075341171\
+    231440736956555270413618581675255342293149119973622969239858152417678164812\
+    112068607
+    """
+
+  /// Python: `hex(int(sys.float_info.max))`
+  static let descriptionFloat64Max_radix16: String =
+    """
+    FFFFFFFFFFFFF80000000000000000000000000000000000000000000000000000000000000\
+    000000000000000000000000000000000000000000000000000000000000000000000000000\
+    000000000000000000000000000000000000000000000000000000000000000000000000000\
+    0000000000000000000000000000000
+    """
+
+  /// Python: `numpy.base_repr(math.factorial(512), base=36)`
+  static let descriptionFactorial512_radix36: String =
+    """
+    7FA5Y7EHR9XHMQ519MBHGYOF8XDYMUX8OZHO9WF1KCM0SSPXV2V45UA73BAFRYM2PFB8CZLTODV\
+    OS3QWA7PYFJ7WAFBI4VF371E27N6XZ4LGWHMFDS4ZH1O3DGNFG4YABUE1G90ORGRTIOGSQVZLSQ\
+    4TKHKHIQ262JVQ0J6LSKAPN5I65AJD33XODVHRNWJ1VSO0Q2FBOUNCPGQG2SFQKR17XHF1OLTV2\
+    MVNJVTDAIYWVJ9ZH7KXT0EPS00IGIVC7MNCU25HFWE37KNMSJQUL5ALUCE5XZVPFQCQGEVEB93B\
+    GA8LKG67PVZ7Q9QMQKIVNIMPT2973MVDTD1D1A0A4QT6NBZYR0TGSZXBV1PD0CHW4SKZJSLBS4Z\
+    W5WCKDY8BCQCE17KKADVLCTVSQL1BZ2PL52DDPB8S5L0ZEG2ZAZF9V4TNJJNO1D9U9JU7B264QZ\
+    5GLHC3Q0Y3BTECGTI8GRENQ2FV4HSEZKPM9OG302KLSY9MBCSOO0FN229AST84TT87LYWOOS71C\
+    54RPJ9RTO9875Z9DE3HPH89EW5I3SV219O04UU09A4KME7DD7IH49ABO79NR4EXFX1VLL4YOHSA\
+    7AHD1LS5YKZ66F4UPG0RCBGG000000000000000000000000000000000000000000000000000\
+    000000000000000000000000000000000000000000000000000000000000000000000000000
+    """
+
+  // MARK: - Basic arithmetic
+
+  func testDivision() {
+    let foo = BigInt("12345678901234567890123456789012345678901234567890123456789012345678901234567890")!
+    let bar = BigInt("351235231535161613134135135135")!
+    let baz = foo / bar
+    XCTAssertEqual(baz, BigInt("35149318157164029155780432046477458820396117503007")!)
```

Comment:
I get a different result with the Python floor-division operator:

```python
foo = 12345678901234567890123456789012345678901234567890123456789012345678901234567890
bar = 351235231535161613134135135135
baz = foo // bar
rem = foo % bar

baz
# 35149318157164029153358504918339691272847595997760

rem
# 15168177833913379478477270290

foo == bar * baz + rem
# True
```

Reply:
That's worrying. I don't know where I got the expected answer from if it wasn't python! So clearly I'm going to have to write some tests with increasingly large divisions until I find out where it breaks, because the old `description` implementation used the division operator and definitely worked! But it was only ever dividing by 10, so probably there's an issue when the divisor is larger than Int.max? I'll figure it out!

Reply:
So I'm still working on figuring this out. Division is unquestionably the most complicated part of this whole thing.

Reply:
Ok, five days later, I have accomplished the fix! It involved re-writing `_findQhat` from scratch. There are now some comments in there pointing to the relevant parts of TAoCP.

Reply:
oops, there's actually still a bug that current tests don't expose. Fix coming soon, along with hopefully a test to verify it.

Reply:
When I add the following to `testDivision()`, there are failures where the `remainder` is negative.

```swift
  func testDivision() {
    for _ in 0 ..< 100 {
      let expected = BigInt(Float64.random(in: 0x1p64 ... 0x1p255))
      let divisor  = BigInt(Float64.random(in: 0x1p64 ... 0x1p128))
      let (quotient, remainder) = expected.quotientAndRemainder(dividingBy: divisor)
      let actual = divisor * quotient + remainder
      XCTAssertEqual(quotient,  expected / divisor)
      XCTAssertEqual(remainder, expected % divisor)
      XCTAssertEqual(
        actual, expected,
        """
        ## FAILURE ##
        ~~~~~~~~~~~~~
              actual: \(actual)
         != expected: \(expected)
        ~~~~~~~~~~~~~
             divisor: \(divisor)
          * quotient: \(quotient)
         + remainder: \(remainder)
        ~~~~~~~~~~~~~
        """)
    }
```

For example:

```text
## FAILURE ##
~~~~~~~~~~~~~
      actual: 34551449761756987820400978628756574120151069926053018399629726511554612953088
 != expected: 34551449761756987820400978628756574120491352292973956863093101118986381164544
~~~~~~~~~~~~~
     divisor: 314741641862945940112684093562865844224
  * quotient: 109777179648832091859813863123394670558
 + remainder: -62781282794539979680630110816414203904
~~~~~~~~~~~~~
```

Reply:
Ok, I'll take a look. It's almost certainly going to be something really simple like the most significant word of the result being larger than `Int.max`.

Reply:
Yup, that's what it was.